redis/src/caching/cache_manager.rs

```diff
@@ -0,0 +1,226 @@
+use crate::caching::statistics::Statistics;
+use crate::caching::{CacheConfig, CacheStatistics};
+use crate::cmd::{CommandCacheInformation, CommandCacheInformationByRef};
+use crate::Value;
+use lru::LruCache;
+use std::cmp::min;
+use std::collections::hash_map::DefaultHasher;
+use std::hash::Hasher;
+use std::num::NonZeroUsize;
+use std::ops::Add;
+use std::sync::Arc;
+use std::time::{Duration, Instant};
+use tokio::sync::watch::{channel, Receiver, Sender};
+
+type RedisCmd = Vec<u8>;
+type RedisKey = Vec<u8>;
+type LRUCacheShard = LruCache<RedisKey, CacheItem>;
+
+struct CacheCmdEntry {
+    cmd: RedisCmd,
+    value: Option<Value>,
+    receiver: Receiver<Value>,
+}
+
+struct CacheItem {
+    ttl: Instant,
+    value_list: Vec<CacheCmdEntry>,
+}
+
+struct ShardedLRU {
+    shards: Vec<std::sync::Mutex<LRUCacheShard>>,
+}
+impl ShardedLRU {
+    const MAX_SHARD_SIZE: usize = 32;
+    fn new(total_key_size: NonZeroUsize) -> Self {
+        // If total cache size is smaller than max shard size then it won't use sharding.
+        let (shard_size, per_lru_size) = if total_key_size.get() >= Self::MAX_SHARD_SIZE {
```

Comment:
is `shard_size` actually `shard_count`? otherwise I don't understand what's the difference between `shard_size` and `per_lru_size`.

Reply:
I've changed it and now it should be clear.